compiler/src/parsing/driver.re

```diff
@@ -2,79 +2,120 @@
 open Lexing;
 open Location;
 
-type error =
-  | AmbiguousParse(list(Parsetree.parsed_program))
-  | NoValidParse;
+let apply_filename_to_lexbuf = (name, lexbuf) => {
+  lexbuf.lex_curr_p = {...lexbuf.lex_curr_p, pos_fname: name};
+  Location.input_name := name;
+};
 
-let report_error = ppf =>
-  fun
-  | AmbiguousParse(parses) =>
-    if (Grain_utils.Config.verbose^
-        || Grain_utils.Config.parser_debug_level^ > 0) {
-      Format.fprintf(
-        ppf,
-        "The Grain compiler had trouble parsing your program. Here were the potential parses:@\n@[<v>%a@]",
-        ppf =>
-          List.iter(x =>
-            Format.fprintf(
-              ppf,
-              "@[%s@]%,",
-              Sexplib.Sexp.to_string_hum @@
-              Parsetree.sexp_of_parsed_program(x),
-            )
-          ),
-        parses,
-      );
-    } else {
-      Format.fprintf(
-        ppf,
-        "The Grain compiler had trouble parsing your program.",
-      );
-    }
-  | NoValidParse =>
-    Format.fprintf(
-      ppf,
-      "The Grain compiler was unable to parse your program. If you see this message, please file an issue at https://github.com/grain-lang/grain",
-    );
+// This parse is very fast, but cannot report useful errors.
+// parse_program_for_syntax_error is 2-5x slower, but can provide proper
+// syntax error information.
+let parse_program = Parser_header.parse_program(Parser.program);
 
-exception Error(Location.t, error);
+module E = MenhirLib.ErrorReports;
+module L = MenhirLib.LexerUtil;
+module I = UnitActionsParser.MenhirInterpreter;
 
-let () =
-  Location.register_error_of_exn(
-    fun
-    | Error(loc, err) =>
-      Some(Location.error_of_printer(loc, report_error, err))
-    | _ => None,
-  );
+let env = checkpoint =>
+  switch (checkpoint) {
+  | I.HandlingError(env) => env
+  | _ => assert(false)
+  };
+
+let state = checkpoint =>
+  switch (I.top(env(checkpoint))) {
+  | Some(I.Element(s, _, _, _)) => I.number(s)
+  | None =>
+    // The parser hasn't left its initial state. The state number is probably
+    // zero, but it's not guaranteed. A future version of Menhir is supposed
+    // to provide a real solution for this.
+    0
+  };
+
+let show = (text, positions) =>
+  E.extract(text, positions) |> E.sanitize |> E.compress |> E.shorten(20);
```

Comment:
What do we want it to be? Why shorten at all?

Reply:
I'm not sure what we want to it be. We can play around with it though... generally we'd want to shorten because I believe this is token-based, and it could get awkward if your token was very long (like a long string, etc.)

Reply:
Alright. I'm fine leaving it as 20 for now. Do you want to open a tracking issue mentioning that it might be adjusted?

Reply:
Do you think it's worth it? I think's probably fine as-is unless we decide that we want to mess around with it.